### DISCUSSION TITLE
Add support for per-test target runners

### DISCUSSION THREAD

**Discussion by sunshowers** [nextest-rs/nextest#2220] [Feature requests]
At Oxide we have a use case where we want to be able to run `pfexec` (`sudo` equivalent) for some tests on illumos. This is generally possible with [target runners](https://nexte.st/book/target-runners) on illumos, but using `pfexec` against every single test is a recipe for a ton of files suddenly being owned by root.

There are hacks around this, but the ideal solution is to support per-test target runners. We should add a way to do that.

This is mostly a matter of adding configuration to `nextest.toml`, since the general infrastructure for target runners can be reused. I'd estimate it would take 3-4 days for me to get it done, including documentation. So it's just a matter of budgeting time for it.

**Comment by sunshowers** [comment]
Some notes from the debugging session today:

* When nextest encounters both `.cargo/config` and `.cargo/config.toml`, it should loudly warn
* within nextest's support for target runners, it should be possible to only use them for the run-phase and not the list phase (or even define separate target runners for the list and run phase)
* the docs should more clearly talk about the non-cross-compile scenario

**Comment by sunshowers** [comment]
One interesting behavior with Cargo's target runners is that `cargo run`, even if invoked from within the target runner, also invokes the target runner. We'll want to probably support this use case by setting `CARGO_TARGET_<...>_RUNNER`, maybe with an optional `recurse = true`.

**Comment by tv42** [comment]
Another use case for target runners is valgrind.

~~`cargo-valgrind` can make to work with nextest with just `CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_RUNNER=cargo-valgrind cargo nextest run`.~~ (Actually, `cargo-valgrind` seems completely unnecessary. It's kinda unclear to me why it actually exists..)

This is how I run valgrind under nextest:

```
CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_RUNNER=valgrind \
  NEXTEST_PROFILE=valgrind \
  cargo nextest run
```

Valgrind is slower, so I also highly recommend something like:

```
[profile.valgrind]
slow-timeout = { period = "5m", terminate-after = 2 }
```

But I'd really like to use `cargo nextest run --profile=valgrind` to set the runner, and not need the shell script wrapper.

(Nextest config `run-extra-args` looked promising, but that's args to the test executable itself, and seems too late for `--config=target.all.runner=cargo-valgrind`)
